### PR TITLE
[Filter/TFLite] remove memcpy after invoke

### LINKED ISSUE #536: [Filter/TF-lite] Memcpy-less tf-lite result output buffering
I checked the address of the output tensor from TF-Lite for checking how much can we believe the output data. The result is, it's not guaranteed.

At below log, each `outputTensor` address is equal.

With @jaeyun-jung @myungjoo , we discussed about this problem and we decided to make the flag, `allocate_in_invoke` of tensorflow-lite, FALSE.
In this case, `tensor_filter` will allocate each memory and buffer will manage it.

```bash

outputTensor address: 0x1a4ba88
** Message: receiving new data [3]
...

outputTensor address: 0x1a4ba88
** Message: receiving new data [4]
...

outputTensor address: 0x1a4ba88
** Message: receiving new data [5]
...

```

### PR DESCRIPTION
**Pull request by helloahn**
remove `memcpy` by changing tflite input/output property dynamic.
it resolves #536 

**Self evaluation:**
1. Build test: [*]Passed [ ]Failed [ ]Skipped
2. Run test: [*]Passed [ ]Failed [ ]Skipped

Signed-off-by: HyoungjooAhn <hello.ahnn@gmail.com>

### CODE REVIEW

File: gst/tensor_filter/tensor_filter_tensorflow_lite_core.cc [hunk 1]
```diff
@@ -103,6 +103,18 @@ TFLiteCore::loadModel ()
       _print_log ("Failed to construct interpreter\n");
       return -2;
     }
+
+    if (interpreter->AllocateTensors () != kTfLiteOk) {
+      printf ("Failed to allocate tensors\n");
```
Comment (jaeyun-jung) [inline gst/tensor_filter/tensor_filter_tensorflow_lite_core.cc]: please change to  '_print_log',  not 'printf'

Comment (jaeyun-jung) [reply gst/tensor_filter/tensor_filter_tensorflow_lite_core.cc]: I think that input-tensor also would be set to dynamic allocation.
Please review the changes I sent.

File: gst/tensor_filter/tensor_filter_tensorflow_lite_core.cc [hunk 2]
```diff
@@ -103,6 +103,26 @@ TFLiteCore::loadModel ()
       _print_log ("Failed to construct interpreter\n");
       return -2;
     }
+
+    /* set allocation type to dynamic for in/out tensors */
+    int tensor_idx;
+
+    int inputTensorSize = interpreter->inputs ().size ();
```
[the remark this answers is not in the document]
Comment (helloahn) [reply gst/tensor_filter/tensor_filter_tensorflow_lite_core.cc]: it's resolved :)

File: gst/tensor_filter/tensor_filter_tensorflow_lite_core.cc [hunk 3]
```diff
@@ -284,30 +304,36 @@ TFLiteCore::invoke (const GstTensorMemory * input, GstTensorMemory * output)
...
     return -3;
   }
 
-  for (int i = 0; i < outputTensorMeta.num_tensors; i++) {
-    int out_index = interpreter->outputs ()[i];
-    TfLiteTensor *output_tensor = interpreter->tensor (out_index);
-
-    g_assert (output_tensor->bytes == output[i].size);
-    memcpy (output[i].data, output_tensor->data.raw, output[i].size);
+  /* if it is not `nullptr`, tensorflow makes `free()` the memory itself. */
```
Comment (leemgs) [inline gst/tensor_filter/tensor_filter_tensorflow_lite_core.cc]: In order to get a well-generated doxygen document, Please /**  ............ */ ASAP.

File: gst/tensor_filter/tensor_filter_tensorflow_lite_core.cc [hunk 4]
```diff
@@ -284,30 +304,36 @@ TFLiteCore::invoke (const GstTensorMemory * input, GstTensorMemory * output)
...
+  for (int i = 0; i < tensors_idx.size (); ++i) {
+    interpreter->tensor (tensors_idx[i])->data.raw = nullptr;
```
Comment (leemgs) [inline gst/tensor_filter/tensor_filter_tensorflow_lite_core.cc]: This approach is better than the exsing technique.

File: gst/tensor_filter/tensor_filter_tensorflow_lite_core.cc [hunk 5]
```diff
@@ -150,7 +170,7 @@ TFLiteCore::setInputTensorProp ()
   auto input_idx_list = interpreter->inputs ();
   inputTensorMeta.num_tensors = input_idx_list.size ();
 
-  for (int i = 0; i < inputTensorMeta.num_tensors; i++) {
+  for (int i = 0; i < inputTensorMeta.num_tensors; ++i) {
```
Comment (leemgs) [inline gst/tensor_filter/tensor_filter_tensorflow_lite_core.cc]: BTW, what is difference between i++ and ++i  in this code?

File: gst/tensor_filter/tensor_filter_tensorflow_lite_core.cc [hunk 6]
```diff
@@ -103,6 +103,26 @@ TFLiteCore::loadModel ()
       _print_log ("Failed to construct interpreter\n");
       return -2;
     }
+
+    /* set allocation type to dynamic for in/out tensors */
```
Comment (leemgs) [inline gst/tensor_filter/tensor_filter_tensorflow_lite_core.cc]: It looks good to me at least.

File: gst/tensor_filter/tensor_filter_tensorflow_lite_core.cc (hunk 3 above)
Comment (helloahn) [reply gst/tensor_filter/tensor_filter_tensorflow_lite_core.cc]: it's applied :)

File: gst/tensor_filter/tensor_filter_tensorflow_lite_core.cc (hunk 5 above)
Comment (helloahn) [reply gst/tensor_filter/tensor_filter_tensorflow_lite_core.cc]: According to the below answer, it looks almost same.
https://stackoverflow.com/a/24887
I just tried to make it better from any type of performance issue.

**Comment by jaeyun-jung** [comment]
@myungjoo Yes, I've tested the changes with Ahn.

**Pull request merged**